fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/SkewShuffleJoinEliminationRule.java

```diff
@@ -0,0 +1,514 @@
+import com.starrocks.sql.optimizer.base.DistributionSpec;
+import com.starrocks.sql.optimizer.base.PhysicalPropertySet;
+import com.starrocks.sql.optimizer.base.RoundRobinDistributionSpec;
+import com.starrocks.sql.optimizer.operator.Operator;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.Projection;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalCTEAnchorOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalDistributionOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalHashAggregateOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalHashJoinOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalJoinOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalMergeOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalSplitConsumeOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalSplitProduceOperator;
+import com.starrocks.sql.optimizer.operator.scalar.BinaryPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CompoundPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.InPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.IsNullPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rewrite.BaseScalarOperatorShuttle;
+import com.starrocks.sql.optimizer.rewrite.ReplaceColumnRefRewriter;
+import com.starrocks.sql.optimizer.rewrite.ScalarOperatorRewriter;
+import com.starrocks.sql.optimizer.rewrite.scalar.ImplicitCastRule;
+import com.starrocks.sql.optimizer.rewrite.scalar.ReduceCastRule;
+import com.starrocks.sql.optimizer.task.TaskContext;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/* rewrite the tree top down
+ * if one shuffle join op is decided as skew, rewrite it as below and do not visit its children
+ *        shuffle join                           union
+ *       /         \     --->            /                   \
+ *  exchange    exchange           shuffle join        broadcast join
+ *     |            |                   /       \           /       \
+ *   child1      child2           exchange  exchange  ex(random)  ex(broadcast)
+ *                                   |            \      /          /
+ *                                   |             \   /          /
+ *                                   |             / \          /
+ *                                   |           /    \        /
+ *                                   |         /       \     /
+ *                                  child1            child2
+ */
```

Comment:
> 1. chid1 and child2's non-skew data are using shuffle join, chid1 and child2's `skew` data are using broadcast join?
> 2. What's the broadcast's distribution type? how to determine its dop?

1. correct
2. for broadcast, left table's skew data will be sent randomly, right table's coresponding data will be sent broadcast. They are pointed by leftSplitConsumerOptForBroadcastJoin and rightSplitConsumerOptForBroadcastJoin's DistributionSpec, the former is RoundRobinDistributionSpec, the latter is DistributionSpec.createReplicatedDistributionSpec(). And PlanFragmentBuilder will translate DistributionSpec into DataPartition in visitPhysicalSplitConsumer, then this  DataPartition  will be stored in splitProduceFragment.getOutputPartitions().add(dataPartition); and Exchang Node to tell exchange sink and source

Reply:
> I see CTE is used in the code. Can you add more infos in the description?

PhysicalCTEAnchorOperator is actually only used for planFragment builder, it will first visit left child which is split producer, then visit right child which is split consumer, actually it has nothing to do with cte.You can see document for details